### ISSUE TITLE
Address the problem of ":"s in the folder names because of the use of identifiers

### ISSUE THREAD

**Issue by swcurran** [bcgov/aries-oca-bundles#22]
The success of using pure DID and Schema/CredDef identifiers as folder paths stops at Windows. This repo can't be cloned on Windows because of ":"s in the folder names.  As such, I'd like to propose a change to how this repository works for Aries Wallets and Agents using it.  Here is my proposal:

- Change the folder structure to the following:
  - `OCABundles/schema/<publisher>/<schema>`, where `<publisher>` and `<schema>` are up to the PR submitter.
  - `OCABundles/credential/<issuer>/<credential>`, where `<issuer>` and `<credential>` are up to the PR submitter.
- The `README.md` ~~or REDIRECT.md~~ file must contain a Markdown table of DID/Identifier pairs that use the OCABundle in the folder. The Markdown table **MUST** have the following columns:

```text
| DID | Identifier | Ledger | URL |
| ---  | --------- | ------ | --- |
|<did>|<identifier> | [<ledger>] | [<URL for object>] |
|<did>|<identifier> |  [<ledger>] | [<URL for object>] |
```

- The table may contain multiple identifiers, and the identifiers may be either Schema or CredDef identifiers.
   - The need for the `REDIRECT.md` file is eliminated with this approach.
   - The `ledger` and `URL` values are optional per row and is for human use only.
      - `ledger` si only for legacy Indy DIDs, and should be in the form `<network>[:<instance>]` as [defined in the `did-indy` specification for the namespace](https://hyperledger.github.io/indy-did-method/#indy-did-method-identifiers) -- e.g., `candy:dev` or `sovrin`.
      - `URL` is a plain (non-Markdown) link to a ledger browser instance of the object.
- A single file in the root folder called `ocabundles.csv` is generated by traversing the `OCABundles` folder and extracting from the markdown table in the `README.md` files the DIDs and Identifiers.
   - The header for the `ocabundles.csv` file is: `"did", "identifier", "OCABundlePath"
   - The <OCABundlePath" excludes the server prefix as the client knows it already.
      - E.g, for this repository, the server prefix is: `https://raw.githubusercontent.com/bcgov/aries-oca-bundles/main/`
   - The `ledger` and `URL` columns from the markdown table are **NOT** included in the file.
   - The `csv` format is proposed because it is the most compact form of the data, but the data could be in JSON if that is easier for the clients and the compactness is not crucial.
- Multiple identifiers using this are accounted for in generating the file, so all identifiers point directly to the relevant OCABundle for a given DID/Identifier combination.
- A GitHub Action will execute on merge to execute the updates to the  `ocabundles.csv` and produce a PR to update the file.

An option to this approach is to either:

- Add two fields `issuer` and `name` to the contents of the file to display where needed, such as in the OCA Explorer.
- Generate a second file `ocabundles-extended.csv` that adds the two fields `issuer` and `name` to the contents of the first file for display where needed. That way the `ocabundles.csv` is kept as short as possible.

Looking for comments/suggestions on this from @amanji @jcdrouin21 @TelegramSam @cvarjao

**Comment by swcurran** [comment]
So @cvarjao you are saying the two step -- load index, load oca-bundle is too much?  Note that the wallet could cache the index, making the load a single step at credential processing time -- at the cost of holding some extra data in the wallet. Also -- it could be a 0 request step if there is no OCA Bundle for the credential -- no need to do anything more than the check the index.

If you look at the current process with the redirect, and with the option of using the CredDef or, if not found, the Schema, so there could be up to 4 requests now -- which also sucks.  My feeling is that just fixing the foldernames is insufficient.

I am a bit worried about the size of the index over time (e.g. 100s of entries), but I think we move have a better longer term solution if that is a concern -- e.g. OCA Bundle on the ledger, returned with the CredDef.

**Comment by swcurran** [comment]
FYI -- I would like to add "PresentationRequests" to this so we can have Presentation Request templates in this repo.

**Comment by swcurran** [comment]
Folks -- revamped the idea per conversations with @cvarjao.  Key changes:

- No redirect file/concept
- Markdown table to hold a list of <DID>, <identifier> pairs that use the OCABundle.

**Comment by swcurran** [comment]
Here is what the CSV would look like if generated on the repo today:

```csv
"did", "identifier", "OCABundlePath"
"63ZiwyeZeazA6AhYRYm2zD", "63ZiwyeZeazA6AhYRYm2zD:2:Member Card:1.5.1", "OCABundles/XUxBrVSALWHLeycAUhrNr9/XUxBrVSALWHLeycAUhrNr9:2:Member Card:1.5.1"
"Trx3R1frdEzbn34Sp1jyX","Trx3R1frdEzbn34Sp1jyX:2:Member Card:1.5.1, "OCABundles/XUxBrVSALWHLeycAUhrNr9/XUxBrVSALWHLeycAUhrNr9:2:Member Card:1.5.1"
"63ZiwyeZeazA6AhYRYm2zD","63ZiwyeZeazA6AhYRYm2zD:2:student_card:1.0, "OCABundles/XUxBrVSALWHLeycAUhrNr9/XUxBrVSALWHLeycAUhrNr9:2:student_card:1.0"
"Ep31SvFAetugFPe5CGzJxt","Ep31SvFAetugFPe5CGzJxt:2:Attestation_numérique_didentité_gouvernemental_(EXP):1.0, "OCABundles/Ep31SvFAetugFPe5CGzJxt/Ep31SvFAetugFPe5CGzJxt:2:Attestation_numérique_didentité_gouvernemental_(EXP):1.0"
"Trx3R1frdEzbn34Sp1jyX","Trx3R1frdEzbn34Sp1jyX:2:student_card:1.0, "OCABundles/XUxBrVSALWHLeycAUhrNr9/XUxBrVSALWHLeycAUhrNr9:2:student_card:1.0"
"XUxBrVSALWHLeycAUhrNr9","XUxBrVSALWHLeycAUhrNr9:2:Member Card:1.5.1, "OCABundles/XUxBrVSALWHLeycAUhrNr9/XUxBrVSALWHLeycAUhrNr9:2:Member Card:1.5.1"
"XUxBrVSALWHLeycAUhrNr9","XUxBrVSALWHLeycAUhrNr9:2:Person:1.0, "OCABundles/XUxBrVSALWHLeycAUhrNr9/XUxBrVSALWHLeycAUhrNr9:2:Person:1.0"
"XUxBrVSALWHLeycAUhrNr9","XUxBrVSALWHLeycAUhrNr9:2:student_card:1.0, "OCABundles/XUxBrVSALWHLeycAUhrNr9/XUxBrVSALWHLeycAUhrNr9:2:student_card:1.0"
"YXCtXE4YhVjULgj5hrk4ML","YXCtXE4YhVjULgj5hrk4ML:2:unverified_person:0.1.0, "OCABundles/YXCtXE4YhVjULgj5hrk4ML/YXCtXE4YhVjULgj5hrk4ML:2:unverified_person:0.1.0"
```

@amanji -- would that work for you in either the wallet or the OCA Explorer? Or would you prefer a less compact representation such as JSON?  And if JSON, please provide an example of the preferred format.

**Comment by amanji** [comment]
My thoughts are the JSON format would be keyed by did as such:

```json
{
    "<did>": [
        {
            "id": "<s_ID or cd_ID>",
            "path": "<path_to_bundle_json>",
            "publisher": "<publisher>",
            "type": "<s or cd>",
            "name": "<name>",
            "description": "<description>"
        }
    ]
}
```

So for example:

```json
{
    "63ZiwyeZeazA6AhYRYm2zD": [
        {
            "id": "63ZiwyeZeazA6AhYRYm2zD:2:Member Card:1.5.1",
            "path": "OCABundles/XUxBrVSALWHLeycAUhrNr9/XUxBrVSALWHLeycAUhrNr9:2:Member Card:1.5.1",
            "publisher": "Law Society of BC",
            "type": "s",
            "name": "Member Card",
            "description": "Member card for LSBC"
        }, {
            "id": "63ZiwyeZeazA6AhYRYm2zD:2:student_card:1.0",
            "path": "OCABundles/XUxBrVSALWHLeycAUhrNr9/XUxBrVSALWHLeycAUhrNr9:2:student_card:1.0",
            "publisher": "BestBC College",
            "type": "s",
            "name": "Student Card",
            "description": "A student card for BestBC College"
        },
        ...
    ],
    "XUxBrVSALWHLeycAUhrNr9": [ ... ],
    ...
    ""
}
```

This should also work for the redirects as proposed

**Comment by swcurran** [comment]
For the file to be loaded by a wallet or other agent, I would not include “publisher”, “type”, “name” or “description”. All of that is in the OCA Bundle and not needed.  We want to minimize the file size to reduce load time and storage needs.  My plan is to have a second file that has a list of the Bundles by Name and Description for use by something like OCA Explorer.

I’ll try to get the arrangement by DID.  It’s a bit tricky to produce the file like that when traversing the file structure, but I think I can do it.

**Comment by cvarjao** [comment]
I was thinking on it being keyed by `schema id` and/or `cred definition id`. That should make for an easy lookup from the wallet (as it it [how it already working](https://github.com/hyperledger/aries-mobile-agent-react-native/blob/main/packages/legacy/core/App/assets/oca-bundles.json)). I am thinking from a quick lookups perfective.
Or is there any advantage to use DID key that I might be missing?

Something like:
```javascript
if (index[schema_id_OR_cred_def_id]) {
   // found a bundle associated with a schema id or cred_def_id
  let bundle = cache[schema_id_OR_cred_def_id]
   if (!bundle) {
       const bundle = await fetch_bundle(schema_id_OR_cred_def_id)
       if (!bundle) throw new Error('Oops. Bundle should be fetchable');
       cache[schema_id_OR_cred_def_id]= bundle
   }
}
```

I am coming from the repository working as a remote caching and the data structures is as close as possible to be readily used, with little to none extra processing.

**Comment by swcurran** [comment]
@cvarjao -- so what would the JSON look like?  I agree we can leave off the DID.

Just:

```
[
    "<id>": "<path>",
    "<id>": "<path>",
]
```

**Comment by cvarjao** [comment]
yes, as json object:
```
{
    "<id>": "<path>",
    "<id>": "<path>",
}
```

Do you think there is any value in the value being an object for future use?
```
{
    "<id>": { path: "<path>" },
    "<id>": { path: "<path>" },
}
```

**Comment by amanji** [comment]
So to be clea,r we would modify the structure above to be:

```json
{
    "<s_ID or cd_ID>": {
            "id": "<s_ID or cd_ID>",
            "path": "<path_to_bundle_json>",
            "publisher": "<publisher>",
            "type": "<\"s\" or \"cd\">",
            "name": "<name>",
            "description": "<description>"
    }
}
```

The `id` property is redundant.

**Comment by swcurran** [comment]
@amanji — we have two separate files. One with just the ID and path, the other with path, publisher, type, name and description.  The first file is for wallets and agents, the second file is for OCA Explorer.

**Comment by amanji** [comment]
Do we need both if that information is present in the JSON for the wallet.

**Comment by swcurran** [comment]
@cvarjao — the one thing I note is the that “id” can have a space in it. Is that an issue for using the “<id>” being the JSON item?

**Comment by swcurran** [comment]
> Do we need both if that information is present in the JSON for the wallet.

The wallet doesn’t need the info that the OCA Explorer needs.  That’s why we have the compact file for the Wallet, and the expanded file for the OCA Explorer.

**Comment by amanji** [comment]
> > Do we need both if that information is present in the JSON for the wallet.
> 
> The wallet doesn’t need the info that the OCA Explorer needs. That’s why we have the compact file for the Wallet, and the expanded file for the OCA Explorer.

I see. I was mistakenly thinking the opposite.

**Comment by amanji** [comment]
I *properly* read your comment above. I see why both are being proposed. Thanks!

**Comment by swcurran** [comment]
When I do just `[ “<id>”: “<path>”, … ]` I get a JSON error — the validator thinks the “:” shouldn’t be there.

I can get this to validate: `[ { “id”: “<id>”, “path”: “<path>”}, …]`

Is that OK?

**Comment by amanji** [comment]
`[ “<id>”: “<path>”, … ]` is not valid JSON since key/values need to be in an object (they are in an array here). The second one is ok.

**Comment by swcurran** [comment]
@amanji — your latest doesn’t have the “<id>” in the JSON — where should that be?

**Comment by amanji** [comment]
> @amanji — your latest doesn’t have the “” in the JSON — where should that be?

What does the "" refer to?

**Comment by swcurran** [comment]
> I **properly** read your comment above. I see why both are being proposed. Thanks!
> 
> I think they should still follow the same structure then:
> 
> Key by "<s_ID or cd_ID>" and value is an object (for the wallet this would just contain the path)

A couple of reasons not to do that:

For the OCA Explorer JSON, the ID is irrelevant as we aren’t using it for anything. I think what we want is to display a list of human readable Publisher+names for the bundles, or perhaps provide a search on the publisher, names, description.  But we don’t want to list them by the obscure IDs.

As well, for the OCA Explorer, we only want the Bundle to be listed once. If we key off the ID, we can get the same Bundle listed multiple times — when it is reused by different identifiers.

**Comment by swcurran** [comment]
@amanji — your latest doesn’t have the “\<id>” in the JSON — where should that be?

**Comment by amanji** [comment]
Assuming `"<id>"` cannot be "" then the object would be: `{ "<id>": { "path": "<path>"} }`

**Comment by amanji** [comment]
> > @amanji — your latest doesn’t have the “” in the JSON — where should that be?
> 
> What does the "" refer to?

Weird the content in the angle braces wasn't showing up.

**Comment by swcurran** [comment]
That gives invalid JSON. Zoom time — see rocketchat.

**Comment by swcurran** [comment]
OK — all figured out.  DIDs removed, JSON set — two files.  Still need to do the documentation, and then I’ll put in the PR.  Sorry for the painful discussion.

**Comment by swcurran** [comment]
Fixed -- woot! #25
